### PR TITLE
RPP RICAP Tensor for HOST and HIP

### PR DESCRIPTION
**Pull request by r-abishek**
Introduces RICAP functionality for HOST and HIP in RPP as per https://arxiv.org/pdf/1811.09030.pdf
Adds corresponding unit/qa/performance tests

### CODE REVIEW

**Comment by kiritigowda** [comment]
@r-abishek this pr has conflicts.

**Comment by kiritigowda** [comment]
@r-abishek this pr has few issues - https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/rpp/pullRequest?prid=12598993

**Comment by r-abishek** [comment]
@kiritigowda Codacy and conflicts resolved

File: src/modules/cpu/kernel/ricap.hpp [hunk 1]
```diff
@@ -0,0 +1,1572 @@
...
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "rppdefs.h"
+#include "rpp_cpu_simd.hpp"
+#include "rpp_cpu_common.hpp"
+
+    //           RICAP output image profile
+    // |------img-roi-1------|------img-roi-2------|
```
Comment (rrawther) [inline src/modules/cpu/kernel/ricap.hpp:28]: how did we decide on this image profile

Comment (r-abishek) [reply src/modules/cpu/kernel/ricap.hpp:28]: Added ricap reference paper link to this comment

File: include/rppdefs.h [hunk 2]
```diff
@@ -80,10 +80,11 @@ typedef enum
     RPP_ERROR_INVALID_DST_CHANNELS      = -8,
     RPP_ERROR_INVALID_SRC_LAYOUT        = -9,
     RPP_ERROR_INVALID_DST_LAYOUT        = -10,
-    RPP_ERROR_INVALID_SRC_DATATYPE     = -11,
-    RPP_ERROR_INVALID_DST_DATATYPE     = -12,
-    RPP_ERROR_INVALID_SRC_OR_DST_DATATYPE      = -13,
-    RPP_ERROR_INSUFFICIENT_DST_BUFFER_LENGTH    = -14
+    RPP_ERROR_INVALID_SRC_DATATYPE      = -11,
+    RPP_ERROR_INVALID_DST_DATATYPE      = -12,
+    RPP_ERROR_INVALID_SRC_OR_DST_DATATYPE     = -13,
```
Comment (rrawther) [inline include/rppdefs.h:85]: Please only have one return code for invalid_datatype

File: include/rppdefs.h [hunk 3]
```diff
@@ -80,10 +80,11 @@ typedef enum
...
+    RPP_ERROR_INSUFFICIENT_DST_BUFFER_LENGTH  = -14,
```
Comment (rrawther) [inline include/rppdefs.h:86]: please rename this to NOT_ENOUGH_MEMORY

File: src/modules/cpu/kernel/ricap.hpp [hunk 4]
```diff
@@ -0,0 +1,1572 @@
...
+    // |------img-roi-1------|------img-roi-2------|
+    // |------img-roi-1------|------img-roi-2------|
+    // |------img-roi-1------|------img-roi-2------|
+    // |------img-roi-1------|------img-roi-2------|
+    // |------img-roi-3------|------img-roi-4------|
+    // |------img-roi-3------|------img-roi-4------|
+    // |------img-roi-3------|------img-roi-4------|
+
+RppStatus ricap_u8_u8_host_tensor(Rpp8u *srcPtr,
```
Comment (rrawther) [inline src/modules/cpu/kernel/ricap.hpp:39]: it is hard to verify the algorithm correctness by reviewing this code.
Is there a unit-test which runs the algorithm in C code  and compares the output with this?

File: src/modules/rppi_validate.hpp [hunk 5]
```diff
@@ -336,4 +336,27 @@ inline void get_dstBatchIndex(rpp::Handle& handle, unsigned int channel, RppiChn
 }
 
 #endif // GPU_SUPPORT
+
+inline int check_roi_out_of_bounds(RpptROIPtr roiPtrImage, RpptDescPtr srcDescPtr, RpptRoiType type)
+{
+    int x, y, w, h;
+    if (type == RpptRoiType::XYWH)
+    {
+        x = (0 <= roiPtrImage->xywhROI.xy.x < srcDescPtr->w) ? roiPtrImage->xywhROI.xy.x : -1;
```
Comment (rrawther) [inline src/modules/rppi_validate.hpp:345]: why is this out of bound values set to -1?

**Review by rrawther — CHANGES_REQUESTED** [review]
please address review comments

File: src/modules/cpu/kernel/ricap.hpp (hunk 4 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/ricap.hpp:39]: Yes Rajy, there's support for QA test and unit-tests to visually verify the outputs.

File: src/modules/rppi_validate.hpp (hunk 5 above)
Comment (snehaa8) [reply src/modules/rppi_validate.hpp:345]: Here we check if point is within bounds as well as check if start offset / dimensions are valid i.e. >0, thus setting this to true / false will only check if value is out of bound and not validate negative x, y, width, height values.

Please correct me if I have understood this incorrectly.

File: src/modules/cpu/kernel/ricap.hpp (hunk 4 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/ricap.hpp:39]: We have compared outputs from both C and assembly codes during development and they match pixelwise.

File: include/rppdefs.h (hunk 2 above)
Comment (r-abishek) [reply include/rppdefs.h]: Done

File: include/rppdefs.h (hunk 3 above)
Comment (r-abishek) [reply include/rppdefs.h]: Done

**Comment by paveltc** [comment]
@r-abishek This test passes both HOST and HIP QA tests. 
I need to talk to you about the unit tests and where to get the inputs for them.

**Comment by paveltc** [comment]
@r-abishek I ran the unit tests for this and sent you an email regarding what I found. If the behavior I saw is expected, this PR should be good to merge.

**Comment by r-abishek** [comment]
@paveltc RICAP is good, its just that the test-suite inputs given to the function are randomized, so creates black areas if inputs are not range bound. 
A bound has been added on the test suite to guarantee a better output, when the demo / unit test is run by users.

**Review by rrawther — APPROVED** [review]
@paveltc to run tests before approving

**Comment by paveltc** [comment]
@rrawther @kiritigowda @r-abishek I am now getting the following error when running the latest version of this PR:
/home/svcbuild/rpp-pr213/rpp/utilities/test_suite/HIP/../rpp_test_suite_common.h:1092:20: error: use of undeclared identifier 'boost'
    randFromDist = boost::math::ibeta_inv(betaParam, betaParam, p); // Computes the inverse of the incomplete beta function on parameter
                   ^
/home/svcbuild/rpp-pr213/rpp/utilities/test_suite/HIP/../rpp_test_suite_common.h:1098:21: error: use of undeclared identifier 'boost'
    randFromDist1 = boost::math::ibeta_inv(betaParam, betaParam, p1);

**Comment by paveltc** [comment]
@r-abishek I was able to successfully run both the HIP and HOST tests and check their output with the original command line we used, the one with the absolute path to the input images.

**Comment by r-abishek** [comment]
@paveltc Sounds good. @kiritigowda I believe RICAP could now be merged.

**Comment by kiritigowda** [comment]
@r-abishek - please resolve conflicts - @paveltc

**Comment by r-abishek** [comment]
@kiritigowda I think it needs a manual trigger? Could you trigger the CI?

**Comment by r-abishek** [comment]
image
@kiritigowda CI with RICAP seems to be working now. Passes

**Comment by r-abishek** [comment]
@sampath1117 @snehaa8 We will need QA passes on tests henceforth. Pls take a look at above. RICAP PR merged into develop is the initial test for this. Audio and Voxel PRs will need to have this.

**Pull request merged**